kimchi/src/snarky/constraint_system.rs

```diff
@@ -158,6 +160,10 @@ pub enum KimchiConstraint<Var, Field> {
     Poseidon {
         state: Vec<Vec<Var>>,
     },
+    Poseidon2 {
+        states: [[Var; 3]; 55],
+        last: [Var; 3],
```

Comment:
these should be constants hardcoded somewhere : o

I see that they're in a generic trait...

```rust
impl SpongeConstants for PlonkSpongeConstantsKimchi {
    const SPONGE_CAPACITY: usize = 1;
    const SPONGE_WIDTH: usize = 3;
    const SPONGE_RATE: usize = 2;
    const PERM_ROUNDS_FULL: usize = 55;
    const PERM_ROUNDS_PARTIAL: usize = 0;
    const PERM_HALF_ROUNDS_FULL: usize = 0;
    const PERM_SBOX: u32 = 7;
    const PERM_FULL_MDS: bool = true;
    const PERM_INITIAL_ARK: bool = false;
}
```

maybe it's a good reason to remove `SpongeConstants` and hardcode everything. For things that still need to change for the legacy sponge, we can pass these values as arguments when we initialize the sponge maybe? But it seems overkill to have a trait for that now if we're only going to implement one thing in snarky

Reply:
> maybe it's a good reason to remove `SpongeConstants` and hardcode everything.

I would prefer if we didn't. Is it so problematic to have
```rust
    Poseidon2 {
        states:
            [
                [Var; PlonkSpongeConstantsKimchi::SPONGE_WIDTH];
                PlonkSpongeConstantsKimchi::PERM_ROUNDS_FULL
            ],
        last: [Var; PlonkSpongeConstantsKimchi::SPONGE_WIDTH],
    }
```
or similar?

Reply:
it's not a problem but then you need to carry that type parameter in a bunch of places, which isn't ideal IMO (I have low tolerance for generics I guess).

BTW I was suggesting this for a future refactor, not for this PR :o

Reply:
I was doing as Matthew's example, I just forgot to replace there. I'll change it now.
For the refactor, in the future we may be able to change the trait to a constant like:
```rust
pub struct SpongeConstants2 {
    sponge_capacity: usize,
    sponge_width: usize,
    sponge_rate: usize,
    perm_rounds_full: usize,
    perm_rounds_partial: usize,
    perm_half_rounds_full: usize,
    perm_sbox: u32,
    perm_full_mds: bool,
    perm_initial_ark: bool,
}
const PlonkSpongeConstantsKimchi: SpongeConstants2 = SpongeConstants2 {
    sponge_capacity: 1,
    sponge_width: 3,
    sponge_rate: 2,
    perm_rounds_full: 55,
    perm_rounds_partial: 0,
    perm_half_rounds_full: 0,
    perm_sbox: 7,
    perm_full_mds: true,
    perm_initial_ark: false,
};
```